vochain/cometbft.go

```diff
@@ -0,0 +1,452 @@
+		"proposer", hex.EncodeToString(req.GetProposerAddress()))
+	return &abcitypes.ResponseFinalizeBlock{
+		AppHash:   root,
+		TxResults: txResults,
+	}, nil
+}
+
+// Commit is the CometBFT implementation of the ABCI Commit method. We currently do nothing here.
+func (app *BaseApplication) Commit(_ context.Context, _ *abcitypes.RequestCommit) (*abcitypes.ResponseCommit, error) {
+	app.prepareProposalLock.Lock()
+	defer app.prepareProposalLock.Unlock()
+	// save state and get hash
+	h, err := app.CommitState()
+	if err != nil {
+		return nil, err
+	}
+	log.Debugw("commit block", "height", app.Height(), "hash", hex.EncodeToString(h))
+	return &abcitypes.ResponseCommit{
+		RetainHeight: 0, // When snapshot sync enabled, we can start to remove old blocks
+	}, nil
+}
+
+// PrepareProposal allows the block proposer to perform application-dependent work in a block before
+// proposing it. This enables, for instance, batch optimizations to a block, which has been empirically
+// demonstrated to be a key component for improved performance. Method PrepareProposal is called every
+// time CometBFT is about to broadcast a Proposal message and validValue is nil. CometBFT gathers
+// outstanding transactions from the mempool, generates a block header, and uses them to create a block
+// to propose. Then, it calls RequestPrepareProposal with the newly created proposal, called raw proposal.
+// The Application can make changes to the raw proposal, such as modifying the set of transactions or the
+// order in which they appear, and returns the (potentially) modified proposal, called prepared proposal in
+// the ResponsePrepareProposal call. The logic modifying the raw proposal MAY be non-deterministic.
+func (app *BaseApplication) PrepareProposal(ctx context.Context,
+	req *abcitypes.RequestPrepareProposal) (*abcitypes.ResponsePrepareProposal, error) {
+	app.prepareProposalLock.Lock()
+	defer app.prepareProposalLock.Unlock()
+	startTime := time.Now()
+
+	type txInfo struct {
+		Data      []byte
+		Addr      *ethcommon.Address
+		Nonce     uint32
+		DecodedTx *vochaintx.Tx
+	}
+	// Rollback the state to discard previous non saved changes
+	app.State.Rollback()
+
+	// Get and execute transactions from the mempool
+	validTxInfos := []txInfo{}
+	for _, tx := range req.GetTxs() {
+		vtx := new(vochaintx.Tx)
+		if err := vtx.Unmarshal(tx, app.ChainID()); err != nil {
+			// invalid transaction
+			log.Warnw("could not unmarshal transaction", "err", err)
+			continue
+		}
+		senderAddr, nonce, err := app.TransactionHandler.ExtractNonceAndSender(vtx)
+		if err != nil {
+			log.Warnw("could not extract nonce and/or sender from transaction", "err", err)
+			continue
+		}
+
+		validTxInfos = append(validTxInfos, txInfo{
+			Data:      tx,
+			Addr:      senderAddr,
+			Nonce:     nonce,
+			DecodedTx: vtx,
+		})
+	}
+
+	// Sort the transactions based on the sender's address and nonce
+	sort.Slice(validTxInfos, func(i, j int) bool {
```

Comment:
FYI now that we depend on Go 1.21 you can use better generics, e.g. https://pkg.go.dev/slices#SortFunc

Reply:
I don't feel ready yet to use Go Generics in a comfortable way. I think I'll let this improvement for the future, or feel free if you want to take it.